applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp

```diff
@@ -0,0 +1,629 @@
+    MatrixType& rLeftHandSideMatrix,
+    VectorType& rRightHandSideVector,
+    ProcessInfo& rCurrentProcessInfo,
+    const bool CalculateStiffnessMatrixFlag,
+    const bool CalculateResidualVectorFlag
+    )
+{
+    KRATOS_ERROR << "You have called to the CalculateAll from the base class for shell elements" << std::endl;
+}
+
+void BaseShellElement::CheckVariables()
+{
+    KRATOS_CHECK_VARIABLE_KEY(DISPLACEMENT);
+    KRATOS_CHECK_VARIABLE_KEY(ROTATION);
+    KRATOS_CHECK_VARIABLE_KEY(VELOCITY);
+    KRATOS_CHECK_VARIABLE_KEY(ANGULAR_VELOCITY);
+    KRATOS_CHECK_VARIABLE_KEY(ACCELERATION);
+    KRATOS_CHECK_VARIABLE_KEY(ANGULAR_ACCELERATION);
+    KRATOS_CHECK_VARIABLE_KEY(VOLUME_ACCELERATION)
+    KRATOS_CHECK_VARIABLE_KEY(DENSITY);
+    KRATOS_CHECK_VARIABLE_KEY(THICKNESS);
+    KRATOS_CHECK_VARIABLE_KEY(SHELL_ORTHOTROPIC_LAYERS);
+    KRATOS_CHECK_VARIABLE_KEY(CONSTITUTIVE_LAW);
+    KRATOS_CHECK_VARIABLE_KEY(SHELL_CROSS_SECTION);
+}
+
+void BaseShellElement::CheckDofs()
+{
+    GeometryType& r_geom = GetGeometry();
+    // verify that the dofs exist
+    for (unsigned int i = 0; i < r_geom.size(); i++)
+    {
+        auto& r_node = r_geom[i];
+        KRATOS_CHECK_VARIABLE_IN_NODAL_DATA(DISPLACEMENT, r_node);
+        KRATOS_CHECK_VARIABLE_IN_NODAL_DATA(ROTATION, r_node);
+
+        KRATOS_CHECK_DOF_IN_NODE(ROTATION_X, r_node);
+        KRATOS_CHECK_DOF_IN_NODE(ROTATION_Y, r_node);
+        KRATOS_CHECK_DOF_IN_NODE(ROTATION_Z, r_node);
+
+        KRATOS_CHECK_DOF_IN_NODE(DISPLACEMENT_X, r_node);
+        KRATOS_CHECK_DOF_IN_NODE(DISPLACEMENT_Y, r_node);
+        KRATOS_CHECK_DOF_IN_NODE(DISPLACEMENT_Z, r_node);
+
+        KRATOS_ERROR_IF(r_node.GetBufferSize() < 2) << "This Element needs "
+            << "at least a buffer size = 2" << std::endl;
+    }
+}
+
+void BaseShellElement::CheckProperties(const ProcessInfo& rCurrentProcessInfo)
+{
+    // check properties
+    if(pGetProperties() == nullptr)
+        KRATOS_ERROR << "Properties not provided for element " << Id() << std::endl;
+
+    const PropertiesType & props = GetProperties();
+
+    const GeometryType& geom = GetGeometry(); // TODO check if this can be const
+
+    if(props.Has(SHELL_CROSS_SECTION)) // if the user specified a cross section ...
+    {
+        const ShellCrossSection::Pointer & section = props[SHELL_CROSS_SECTION];
+        if(section == nullptr)
+            KRATOS_ERROR << "SHELL_CROSS_SECTION not provided for element " << Id() << std::endl;
+
+        section->Check(props, geom, rCurrentProcessInfo);
+    }
+    else if (props.Has(SHELL_ORTHOTROPIC_LAYERS))
+    {
+        CheckSpecificProperties();
+
+        // perform detailed orthotropic check later in shell_cross_section
+    }
+    else // ... allow the automatic creation of a homogeneous section from a material and a thickness
+    {
+        CheckSpecificProperties();
+
+        ShellCrossSection::Pointer dummySection = ShellCrossSection::Pointer(new ShellCrossSection());
```

Comment:
@RiccardoRossi is not very fan of the new for shared_pointers

Reply:
question: why do you need to make it a pointer? (if you do, please use Kratos::make_shared)